File: leetcode/cpp/medium_algorithms.cpp
/**
 * @file medium_algorithms.cpp
 * @brief medium_algorithms implementation.
 */

#include "medium_algorithms.h"

namespace medium_algorithms {

/**
 * Constructor
 */
MediumAlgorithms::MediumAlgorithms() {}

/**
 * Destructor
 */
MediumAlgorithms::~MediumAlgorithms() {}

/**
 * Initialize
 */
bool MediumAlgorithms::init() {
	return true;
}

/**
 * Solution for the problem 2 Add_Two_Numbers. * 
 */
ListNode* MediumAlgorithms::AddTwoNumbersS2(ListNode* l1, ListNode* l2) {
  ListNode* result = new ListNode(0);
  ListNode* p_result = result;
  int carry = 0;

  while (l1 || l2 || carry) {
    int d1 = l1 ? l1->val : 0;
    int d2 = l2 ? l2->val : 0;
    int sum = d1 + d2 + carry;
    carry = sum / 10;
    p_result->next = new ListNode(sum % 10);
    p_result = p_result->next;
    l1 = l1 ? l1->next : nullptr;
    l2 = l2 ? l2->next : nullptr;
  }  

  return result->next;
}

}  // namespace easy

File: leetcode/python/medium_algorithms.py
"""
  @file medium_algorithms.py
  @brief Contains medium algorithms from LeetCode.
"""

from common import *


class MediumAlgorithms(object):
  """
    @class MediumAlgorithms
    @brief Contains medim algorithms from LeetCode.
  """
  def __init__(self):
    pass


  def AddTwoNunmberS2(self, l1: ListNode, l2: ListNode) -> ListNode:
    """
    @brief You are given two non-empty linked lists representing two non-negative integers.
    The digits are stored in reverse order and each of their nodes contain a single digit.
    Add the two numbers and return it as a linked list.

    Args:
      l1 (ListNode): First linked list.
      l2 (ListNode): Second linked list.
    Returns:
      (ListNode): Sum of two linked lists.

    @note You may assume the two numbers do not contain any leading zero, except the number 0 itself.
    @example
    l1 = [2, 4, 3]
    l2 = [5, 6, 4]
    result = AddTwoNunmberS2(l1, l2)
    # Result = 342 + 465 = 807 = [7, 0, 8]
    """
    result = ListNode(0)
    curr = result
    carry = 0

    while l1 or l2 or carry:
      if l1:
        carry += l1.val
        l1 = l1.next
      if l2:
        carry += l2.val
        l2 = l2.next
      curr.next = ListNode(carry % 10)
      curr = curr.next
      carry //= 10

    return result.next


if __name__ == '__main__':
  sol = MediumAlgorithms()

  # Test AddTwoNunmberS2
  l1 = ListNode(2, ListNode(4, ListNode(2)))
  l2 = ListNode(5, ListNode(6, ListNode(4)))
  expected_result = ListNode(7, ListNode(0, ListNode(8)))
  actual_result = sol.AddTwoNunmberS2(l1, l2)
  # Assert that each node in the linked list is equal
  while actual_result and expected_result:
    assert actual_result.val == expected_result.val, f'AddTwoNunmberS2 - test failed !'
    actual_result = actual_result.next
    expected_result = expected_result.next
    
  
  print('All tests passed')